File: sample/src/main/java/com/readdle/swiftjava/sample/SampleReference.java
package com.readdle.swiftjava.sample;

import com.readdle.codegen.anotation.SwiftBlock;
import com.readdle.codegen.anotation.SwiftCallbackFunc;
import com.readdle.codegen.anotation.SwiftDelegate;
import com.readdle.codegen.anotation.SwiftError;
import com.readdle.codegen.anotation.SwiftFunc;
import com.readdle.codegen.anotation.SwiftGetter;
import com.readdle.codegen.anotation.SwiftReference;
import com.readdle.codegen.anotation.SwiftSetter;
import com.readdle.swiftjava.sample.asbtracthierarhy.AbstractType;
import com.readdle.swiftjava.sample.asbtracthierarhy.FirstChild;
import com.readdle.swiftjava.sample.asbtracthierarhy.FourthChild;
import com.readdle.swiftjava.sample.asbtracthierarhy.SecondChild;
import com.readdle.swiftjava.sample.asbtracthierarhy.ThirdChild;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;

@SwiftReference
public class SampleReference {

    @SwiftReference
    public static class SampleReferenceEnclose {


        // Swift JNI private native pointer
        private long nativePointer = 0L;

        // Swift JNI private constructor
        // Should be private. Don't call this constructor from Java!
        private SampleReferenceEnclose() {
        }

        // Swift JNI release method
        public native void release();

        @NonNull
        public static native SampleReferenceEnclose init();
    }

    @SwiftDelegate(protocols = {"SampleBlockDelegate"})
    public interface SampleInterfaceDelegateAndroid {

        @SwiftCallbackFunc
        void onCall(@NonNull Integer pr1, @NonNull Integer pr2, @NonNull Double pr3, @NonNull Double pr4);

    }

    // Swift JNI private native pointer
    private long nativePointer = 0L;

    // Swift JNI private constructor
    // Should be private. Don't call this constructor from Java!
    private SampleReference() {
    }

    // Swift JNI release method
    public native void release();

    @NonNull
    public native SampleValue getRandomValue();

    public native void saveValue(@NonNull SampleValue value);

    public native void funcThrows() throws SwiftError;

    @Nullable
    public native SampleReference funcWithNil();

    @NonNull
    public static native SampleReference init();

    // TODO: Impossible to generate for now. Add extra check for JavaSwift protocol before casting to .javaObject()
    //@Nullable @SwiftFunc
    //public native SampleDelegateAndroid getDelegate();

    @SwiftSetter
    public native void setDelegate(@Nullable SampleDelegateAndroid delegate);

    @NonNull
    public native Long tick();

    public native void tickWithBlock(@NonNull SampleInterfaceDelegateAndroid delegate);

    @NonNull
    public native Float floatCheck(@NonNull Float var1);

    @NonNull
    public native Double doubleCheck(@NonNull Double var1);

    @NonNull
    public native Exception exceptionCheck(@NonNull Exception var1);

    @NonNull
    public native SampleReferenceEnclose enclose(@NonNull SampleReferenceEnclose var1);

    @SwiftGetter @NonNull
    public native String getString();

    @SwiftSetter
    public native void setString(@NonNull String string);

    @SwiftGetter("staticString") @NonNull
    public native static String getStaticString();

    @SwiftSetter("staticString")
    public native static void setStaticString(@NonNull String string);

    @SwiftBlock("(Error?, String?) -> String?")
    public interface CompletionBlock {
        @Nullable
        String call(@Nullable Exception error, @Nullable String string);
    }

    @SwiftFunc("funcWithBlock(completion:)") @Nullable
    public native String funcWithBlock(@SwiftBlock CompletionBlock block);

    @NonNull
    public native List<AbstractType> getAbstractTypeList();

    @NonNull
    public native AbstractType getAbstractType();

    @NonNull
    public native FirstChild getFirstChild();

    @NonNull
    public native SecondChild getSecondChild();

    @NonNull
    public native ThirdChild getThirdChild();

    @NonNull
    public native FourthChild getFourthChild();

    public native void throwableFunc(@NonNull SampleDelegateAndroid delegate, @NonNull Boolean flag) throws Exception;

    @NonNull
    public native String throwableFuncWithReturnType(@NonNull SampleDelegateAndroid delegate, @NonNull Boolean flag) throws Exception;

}

File: sample/src/main/java/com/readdle/swiftjava/sample/SampleReference.kt
package com.readdle.swiftjava.sample

import com.readdle.codegen.anotation.*
import com.readdle.swiftjava.sample.asbtracthierarhy.*

@SwiftReference
class SampleReference private constructor() {

    @SwiftReference
    class SampleReferenceEnclose private constructor() {

        // Swift JNI private native pointer
        private val nativePointer = 0L

        // Swift JNI release method
        external fun release()

        companion object {
            @JvmStatic
            external fun init(): SampleReferenceEnclose
        }
    }

    @SwiftDelegate(protocols = ["SampleBlockDelegate"])
    interface SampleInterfaceDelegateAndroid {
        @SwiftCallbackFunc
        fun onCall(pr1: Int, pr2: Int, pr3: Double, pr4: Double)
    }

    // Swift JNI private native pointer
    private val nativePointer = 0L

    // Swift JNI release method
    external fun release()

    val randomValue: SampleValue
        external get

    external fun saveValue(value: SampleValue)

    @Throws(SwiftError::class)
    external fun funcThrows()

    external fun funcWithNil(): SampleReference?

    // TODO: Impossible to generate for now. Add extra check for JavaSwift protocol before casting to .javaObject()
    //@Nullable @SwiftFunc
    //public native SampleDelegateAndroid getDelegate();
    @SwiftSetter
    external fun setDelegate(delegate: SampleDelegateAndroid)

    external fun tick(): Long

    external fun tickWithBlock(delegate: SampleInterfaceDelegateAndroid)

    external fun floatCheck(var1: Float): Float

    external fun doubleCheck(var1: Double): Double

    external fun exceptionCheck(var1: Exception): Exception

    external fun enclose(var1: SampleReferenceEnclose): SampleReferenceEnclose

    @get:SwiftGetter
    @set:SwiftSetter
    var string: String
        external get
        external set

    @SwiftBlock("(Error?, String?) -> String?")
    interface CompletionBlock {
        fun call(error: Exception?, string: String?): String?
    }

    @SwiftFunc("funcWithBlock(completion:)")
    external fun funcWithBlock(@SwiftBlock block: CompletionBlock?): String?

    val abstractTypeList: List<AbstractType>
        external get

    @get:SwiftFunc("getAbstractType()")
    val abstractType: AbstractType
        external get

    val firstChild: FirstChild
        external get

    val secondChild: SecondChild
        external get

    val thirdChild: ThirdChild
        external get

    val fourthChild: FourthChild
        external get

    @Throws(Exception::class)
    external fun throwableFunc(delegate: SampleDelegateAndroid, flag: Boolean)

    @Throws(Exception::class)
    external fun throwableFuncWithReturnType(delegate: SampleDelegateAndroid, flag: Boolean): String

    val randomCustomSampleValue: CustomSampleValue
        external get

    external fun saveCustomSampleValue(value: CustomSampleValue)

    external fun oneMoreReferenceTableOverflow(delegate: SampleDelegateAndroid)

    companion object {

        @JvmStatic
        external fun init(): SampleReference

        @JvmStatic @SwiftGetter("staticString")
        external fun getStaticString(): String

        @JvmStatic @SwiftSetter("staticString")
        external fun setStaticString(value: String)

    }
}